rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+increasingly unweidly, a new context structure will be introduced to
+carry the data.
+
+```rust
+struct SourceContext<'a> {
+    name: &'a str,
+    identifier: Box<str>,
+    globals: &'a GlobalOptions,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+}
+
+trait SourceConfig {
+    async fn build(&self, context: SourceContext<'_>) -> crate::Result<sources::Source>;
+}
+```
+
+### Sink configuration
+
+A new global configuration setting will be added to all sources,
+flagging one or more sinks as "authoritative". This setting will change
+the behavior of finalization as described above. When an event is
+delivered to an authoritative sink, its finalization status is
+immediately delivered to all sources of the event. In order to prevent
+another status from being delivered at a later point, the status of the
+event is changed to "no-op", the same as if it was merged into another
+event.
+
+```rust
+struct SinkOuter {
+    // … existing fields …
+
+    #[serde(default)]
+    authoritative: bool,
+}
+```
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+# Global enable option
+acknowledgements = true
+
+[sources.my_source_id]
+  # Per-source enable option
+  acknowledgements = true
+```
+
+We would also need to document when acknowledgement happens for each source.
+
+## Rationale
+
+The above structure provides for several considerations:
+
+1.  This the minimum amount of data that can be added to the metadata to
+    fully support this feature, amounting to a single shared reference
+    as the `Option` is optimized into the `Arc`.
+
+2.  The use of `Arc` reference counting for finalization prevents events
+    from "escaping" without providing a status indication.
+
+3.  If a source does not need or is not configured to require
+    finalization, it will not contribute to the list of sources and so
+    has no additional event overhead, and no additional allocations when
+    the event is created.
+
+4.  Sending the finalization status to the source does not require any
+    lookups or topology traversal.
+
+5.  No additional work is required to handle dropped sources due to
+    topology reconfiguration, other than the expected checking for a
+    closed channel when sending.
+
+## Drawbacks
+
+1.  This adds a base size overhead to each event, even for
+    configurations that do not support or require end-to-end
+    acknowledgement.
+
+## Alternatives
+
+1.  The set of sources could be stored in a more customary `Vec`. This
+    provides for merging multiple sources with a minimum of
+    code. However, merged events already have other overhead, and it
+    increases the data required for this array by an additional word.
+
+2.  The source could be stored as simply the unique identifier
+    string. This requires that all reporting of finalization status
+    proceed through a dictionary lookup instead of simply sending it
+    through a channel, increasing the run-time overhead.
```

Comment:
I dunno, this might have legs. My one major concern with this RFC's proposal is increase in `Event` size. That concern goes away in the future if we we take these constant costs and share them across a large batch, of course, but I worry about this today. Similar concern for how many tag duplicates we have in-memory that could be shared across a batch. I'd feel less concerned about the size bump here if we were sharing this across 1024 `Event`s and no per `Event`, I guess is what I'm getting at. 

It feels like we're modeling lifetimes here in some way. The source has ownership of some `Event` and we're passing it down the topology, keeping a weak reference back up at the source and strong references downstream. That's complicated by merging events and reloads even more but there's some bug in my mind that we could lean more onto the type system somehow. If we didn't reload a source could pass an ack closure into an `Event` it creates that is called on `Drop`. I don't know what to do about disk serialization or reloads for this method, but we would reduce some of the size of the event metadata and miss out on branching inherent in a one-shot that may or may not be closed.

Reply:
With the current proposal, the event itself increases by a single word (the pointer in `Option<Arc<EventFinalizer>>`. However, there will be one `EventFinalizer` allocated for each source event, plus one `BatchNotifier` for each batch of events. Those sies don't change if we store the data in a separate table, so it becomes a matter of direct access on the heap with higher costs to allocate and free, or in a table with higher costs to access.

I agree there is a lifetime element to this proposal. I'm definitely open to a solution that leverages the type system, but merging, fanout, and buffering/reload make that complicated, combined with the fact it has to be a non-generic type (AFAICT at least).

Reply:
> I agree there is a lifetime element to this proposal. I'm definitely open to a solution that leverages the type system, but merging, fanout, and buffering/reload make that complicated, combined with the fact it has to be a non-generic type (AFAICT at least).

As far as I can tell as well. It _feels_ like there ought to be something but I agree that it's real likely if we start going down generics we'll run into a path that leads to madness.